File: validation/dma/test_dma.py
"""DMA validation: single-transfer completion + error handling.

Burst/chained modes deferred.
"""
from __future__ import annotations

SRAM_BASE = 0x10000000
DMA_BASE = 0x20002000
SRC = DMA_BASE + 0x00
DST = DMA_BASE + 0x04
LEN = DMA_BASE + 0x08
CTRL = DMA_BASE + 0x0C
STATUS = DMA_BASE + 0x10
IRQ_CLEAR = DMA_BASE + 0x14
ERR_CODE = DMA_BASE + 0x18

INTC_BASE = 0x20003000
INTC_ENABLE = INTC_BASE + 0x00
INTC_PENDING = INTC_BASE + 0x04
INTC_ACK = INTC_BASE + 0x08
INTC_CLEAR = INTC_BASE + 0x0C

IRQ_DMA = 2

BUSY = 1 << 0
DONE = 1 << 1
ERROR = 1 << 2


def _pattern(nwords: int) -> list[int]:
    return [0xA5000000 | i for i in range(nwords)]


def test_dma_completes(soc):
    src, dst, nwords = SRAM_BASE, SRAM_BASE + 0x1000, 16
    words = _pattern(nwords)
    for i, w in enumerate(words):
        soc.sram_write_word(src + i * 4, w)
    soc.write(INTC_ENABLE, 1 << IRQ_DMA)
    soc.write(SRC, src)
    soc.write(DST, dst)
    soc.write(LEN, nwords * 4)
    soc.write(CTRL, 0x1 | 0x2)  # START | IRQ_ENABLE
    soc.run_until(lambda: soc.read(STATUS) & (DONE | ERROR), max_ticks=1000)
    assert soc.read(STATUS) & DONE
    assert not (soc.read(STATUS) & ERROR)
    for i, w in enumerate(words):
        assert soc.sram_read_word(dst + i * 4) == w
    assert soc.read(ERR_CODE) == 0
    assert soc.read(INTC_PENDING) & (1 << IRQ_DMA)
    assert soc.read(INTC_ACK) == IRQ_DMA
    soc.write(INTC_CLEAR, IRQ_DMA)
    soc.write(IRQ_CLEAR, 1)
    # PERF counted real bytes.
    assert soc.read(0x20004008) == nwords * 4  # PERF DMA_BYTES


def test_dma_invalid_address(soc):
    soc.write(SRC, 0x30000000)
    soc.write(DST, SRAM_BASE)
    soc.write(LEN, 16)
    soc.write(CTRL, 0x1)
    soc.step(2)
    assert soc.read(STATUS) & ERROR
    assert soc.read(ERR_CODE) == 1


def test_dma_misaligned(soc):
    soc.write(SRC, SRAM_BASE + 1)
    soc.write(DST, SRAM_BASE + 0x1000)
    soc.write(LEN, 16)
    soc.write(CTRL, 0x1)
    soc.step(2)
    assert soc.read(STATUS) & ERROR
    assert soc.read(ERR_CODE) == 2


def test_dma_bad_length(soc):
    for bad in (0, 3, 6):
        soc.write(SRC, SRAM_BASE)
        soc.write(DST, SRAM_BASE + 0x1000)
        soc.write(LEN, bad)
        soc.write(CTRL, 0x1)
        soc.step(2)
        assert soc.read(STATUS) & ERROR, f"len={bad}"
        assert soc.read(ERR_CODE) == 3
        soc.write(IRQ_CLEAR, 1)


def test_dma_start_while_busy_ignored(soc):
    src, dst = SRAM_BASE, SRAM_BASE + 0x2000
    soc.write(SRC, src)
    soc.write(DST, dst)
    soc.write(LEN, 64)
    soc.write(CTRL, 0x1)
    assert soc.read(STATUS) & BUSY
    stalls_before = soc.read(0x20004010)  # PERF STALLS
    soc.write(CTRL, 0x1)  # second START while busy
    assert soc.read(STATUS) & BUSY
    assert soc.read(0x20004010) >= stalls_before + 1
    soc.run_until(lambda: soc.read(STATUS) & DONE, max_ticks=1000)

File: soc_c/tests/test_dma.c
/* DMA validation in C. Ports validation/dma/test_dma.py. */
#include <assert.h>
#include <stdio.h>

#include "soc.h"

#define SRC (SOC_DMA_BASE + 0x00)
#define DST (SOC_DMA_BASE + 0x04)
#define LEN (SOC_DMA_BASE + 0x08)
#define CTRL (SOC_DMA_BASE + 0x0C)
#define STATUS (SOC_DMA_BASE + 0x10)
#define IRQ_CLEAR (SOC_DMA_BASE + 0x14)
#define ERR_CODE (SOC_DMA_BASE + 0x18)

#define INTC_ENABLE (SOC_INTC_BASE + 0x00)
#define INTC_PENDING (SOC_INTC_BASE + 0x04)
#define INTC_ACK (SOC_INTC_BASE + 0x08)
#define INTC_CLEAR (SOC_INTC_BASE + 0x0C)

#define IRQ_DMA 2
#define BUSY (1U << 0)
#define DONE (1U << 1)
#define ERROR (1U << 2)

typedef struct {
    soc_t *s;
    uint32_t mask;
} pred_ctx_t;

static int pred_done_or_error(void *ctx) {
    pred_ctx_t *p = (pred_ctx_t *)ctx;
    uint32_t st = 0;
    soc_read(p->s, STATUS, &st);
    return (st & (DONE | ERROR)) ? 1 : 0;
}

int main(void) {
    soc_t s;
    soc_config_t cfg;
    uint32_t v = 0;
    int i;

    soc_config_default(&cfg);
    soc_init(&s, &cfg);
    soc_boot(&s, NULL, 0);

    /* Completes: 16 words SRAM->SRAM with IRQ. */
    {
        uint32_t src = SOC_SRAM_BASE, dst = SOC_SRAM_BASE + 0x1000;
        pred_ctx_t pc = {&s, DONE | ERROR};
        for (i = 0; i < 16; i++)
            assert(soc_sram_write_word(&s, src + i * 4, 0xA5000000U | (uint32_t)i) == SOC_OK);
        assert(soc_write(&s, INTC_ENABLE, 1U << IRQ_DMA) == SOC_OK);
        assert(soc_write(&s, SRC, src) == SOC_OK);
        assert(soc_write(&s, DST, dst) == SOC_OK);
        assert(soc_write(&s, LEN, 16 * 4) == SOC_OK);
        assert(soc_write(&s, CTRL, 0x3) == SOC_OK);
        assert(soc_run_until(&s, pred_done_or_error, &pc, 1000) >= 0);
        assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & DONE));
        assert(!(v & ERROR));
        for (i = 0; i < 16; i++) {
            assert(soc_sram_read_word(&s, dst + i * 4, &v) == SOC_OK &&
                   v == (0xA5000000U | (uint32_t)i));
        }
        assert(soc_read(&s, ERR_CODE, &v) == SOC_OK && v == 0);
        assert(soc_read(&s, INTC_PENDING, &v) == SOC_OK && (v & (1U << IRQ_DMA)));
        assert(soc_read(&s, INTC_ACK, &v) == SOC_OK && v == IRQ_DMA);
        assert(soc_write(&s, INTC_CLEAR, IRQ_DMA) == SOC_OK);
        assert(soc_write(&s, IRQ_CLEAR, 1) == SOC_OK);
        assert(soc_read(&s, SOC_PERF_BASE + 0x08, &v) == SOC_OK && v == 16 * 4);
    }

    /* Invalid address -> ERR_ADDR=1. */
    soc_boot(&s, NULL, 0);
    assert(soc_write(&s, SRC, 0x30000000U) == SOC_OK);
    assert(soc_write(&s, DST, SOC_SRAM_BASE) == SOC_OK);
    assert(soc_write(&s, LEN, 16) == SOC_OK);
    assert(soc_write(&s, CTRL, 0x1) == SOC_OK);
    soc_step(&s, 2);
    assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & ERROR));
    assert(soc_read(&s, ERR_CODE, &v) == SOC_OK && v == 1);

    /* Misaligned -> ERR_ALIGN=2. */
    soc_boot(&s, NULL, 0);
    assert(soc_write(&s, SRC, SOC_SRAM_BASE + 1) == SOC_OK);
    assert(soc_write(&s, DST, SOC_SRAM_BASE + 0x1000) == SOC_OK);
    assert(soc_write(&s, LEN, 16) == SOC_OK);
    assert(soc_write(&s, CTRL, 0x1) == SOC_OK);
    soc_step(&s, 2);
    assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & ERROR));
    assert(soc_read(&s, ERR_CODE, &v) == SOC_OK && v == 2);

    /* Bad lengths -> ERR_LEN=3. */
    {
        uint32_t bads[3] = {0, 3, 6};
        for (i = 0; i < 3; i++) {
            soc_boot(&s, NULL, 0);
            assert(soc_write(&s, SRC, SOC_SRAM_BASE) == SOC_OK);
            assert(soc_write(&s, DST, SOC_SRAM_BASE + 0x1000) == SOC_OK);
            assert(soc_write(&s, LEN, bads[i]) == SOC_OK);
            assert(soc_write(&s, CTRL, 0x1) == SOC_OK);
            soc_step(&s, 2);
            assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & ERROR));
            assert(soc_read(&s, ERR_CODE, &v) == SOC_OK && v == 3);
            assert(soc_write(&s, IRQ_CLEAR, 1) == SOC_OK);
        }
    }

    /* START while busy ignored + stall counted. */
    {
        uint32_t stalls_before = 0, stalls_after = 0;
        soc_boot(&s, NULL, 0);
        assert(soc_write(&s, SRC, SOC_SRAM_BASE) == SOC_OK);
        assert(soc_write(&s, DST, SOC_SRAM_BASE + 0x2000) == SOC_OK);
        assert(soc_write(&s, LEN, 64) == SOC_OK);
        assert(soc_write(&s, CTRL, 0x1) == SOC_OK);
        assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & BUSY));
        assert(soc_read(&s, SOC_PERF_BASE + 0x10, &stalls_before) == SOC_OK);
        assert(soc_write(&s, CTRL, 0x1) == SOC_OK);
        assert(soc_read(&s, STATUS, &v) == SOC_OK && (v & BUSY));
        assert(soc_read(&s, SOC_PERF_BASE + 0x10, &stalls_after) == SOC_OK &&
               stalls_after >= stalls_before + 1);
    }

    printf("DMA OK\n");
    return 0;
}
